src/Mod/Part/Gui/TaskAttacher.cpp

```diff
@@ -345,47 +350,194 @@ bool TaskAttacher::updatePreview()
+                "Unsuitable selection: '%s' cannot be attached to '%s' from within it's group '%s'.\n",
+                attachingObj->getFullLabel(), subObj->getFullLabel(), group->getFullLabel());
+            rootObj = nullptr;
+            return;
+        }
+
+        // In case the attaching object is in a link to a part.
+        // For instance : 
+        // - Part1
+        // - - LinkToPart2
+        // - - - Cube
+        // - - - Sketch
+        obj = obj->getLinkedObject();
+
+        if (obj == group) {
+            ++i;
+            obj = doc->getObject(names[i].c_str());
+            if (!obj) {
+                return;
+            }
+
+            rootObj = obj;
+
+            // Rebuild 'sub' starting from the next element after the current 'name'
+            sub = "";
+            for (size_t j = i + 1; j < names.size(); ++j) {
+                sub += names[j];
+                if (j != names.size() - 1) {
+                    sub += ".";  // Add a period between elements
+                }
+            }
+            return;
+        }
+    }
+
+    // if we reach this point it means that attaching object's group is outside of
+    // the scope of the attached object. For instance: 
+    // - Part1
+    // - - Part2
+    // - - - Cube
+    // - - Part3
+    // - - - Sketch
+    // In this case the selection is not acceptable.
+    rootObj = nullptr;
+}
+
+void TaskAttacher::handleInitialSelection()
+{
+    // We handle initial selection only if it is not attached yet.
+    App::DocumentObject* obj = ViewProvider->getObject();
+    Part::AttachExtension* pcAttach = obj->getExtensionByType<Part::AttachExtension>();
+    std::vector<App::DocumentObject*> refs = pcAttach->AttachmentSupport.getValues();
+
+    if (!refs.empty()) {
+        return;
+    }
+    std::vector<std::string> objNames;
+    std::vector<std::string> subNames;
+
+    auto sel = Gui::Selection().getSelectionEx("",
+        App::DocumentObject::getClassTypeId(), Gui::ResolveMode::NoResolve);
+    for (auto& selObj : sel) {
+        std::vector<std::string> subs = selObj.getSubNames();
+        const char* objName = selObj.getFeatName();
+        for (auto& sub : subs) {
+            objNames.push_back(objName);
+            subNames.push_back(sub);
+        }
     }
+    addToReference(objNames, subNames);
 }
 
 void TaskAttacher::onSelectionChanged(const Gui::SelectionChanges& msg)
 {
-    if (!ViewProvider)
+    if (!ViewProvider) {
         return;
+    }
 
     if (msg.Type == Gui::SelectionChanges::AddSelection) {
-        if (iActiveRef < 0)
-            return;
+        addToReference(msg.pObjectName, msg.pSubName);
+    }
+}
 
-        // Note: The validity checking has already been done in ReferenceSelection.cpp
-        Part::AttachExtension* pcAttach = ViewProvider->getObject()->getExtensionByType<Part::AttachExtension>();
+void TaskAttacher::addToReference(std::vector<std::string> objNames, std::vector<std::string> subNames)

```

Comment:
Do we need to copy `objNames` and `subNames`?

Reply:
no. Changing to 
`addToReference(const std::vector<std::string>& objNames, const std::vector<std::string>& subNames)`